File: tictactoe.py
import random
import time


# default wait time (to ensure smooth gameplay)
DELAY = 2


def show_board(board):
    """Display well-formatted board.
    @type board: list
    @rtype: None
    """
    print("-------------")
    print(f"| {board[0]} | {board[1]} | {board[2]} |")
    print("|---+---+---|")
    print(f"| {board[3]} | {board[4]} | {board[5]} |")
    print("|---+---+---|")
    print(f"| {board[6]} | {board[7]} | {board[8]} |")
    print("-------------\n\n")


def get_p_move(board):
    """Get and validate user move based on current board state.
    @type board: list
    @rtype: int
    """
    available = []

    for i in range(len(board)):
        if board[i] != "X" and board[i] != "O":
            available.append(i)

    while True:
        try:
            move = int(input("Your move: "))
            print()
        except ValueError:
            print("Invalid entry.")
            continue

        if not((move - 1) in available):
            print("That spot is unavailable.")
            continue
        else:
            break
    
    return move - 1


def check_for_win(board):
    """Check board for whether someone has won yet.
    @type board: list
    @rtype: None
    """

    # check if cpu wins
    if board[0] == board[1] == board[2] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[3] == board[4] == board[5] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[6] == board[7] == board[8] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[0] == board[3] == board[6] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[1] == board[4] == board[7] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[2] == board[5] == board[8] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[0] == board[4] == board[8] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()
    elif board[2] == board[4] == board[6] == "X":
        print("CPU WINS!")
        time.sleep(DELAY)
        exit()

    # check if player wins
    elif board[0] == board[1] == board[2] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[3] == board[4] == board[5] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[6] == board[7] == board[8] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[0] == board[3] == board[6] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[1] == board[4] == board[7] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[2] == board[5] == board[8] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[0] == board[4] == board[8] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    elif board[2] == board[4] == board[6] == "O":
        print("YOU WIN!")
        time.sleep(DELAY)
        exit()
    else:
        pass


def cpu_strategic_move(board):
    """Determine optimal cpu move based on current state of board.
    @type board: list
    @rtype: int
    """

    # check for spots where cpu is one move away from winning
    if (board[0] == board[1] == "X") and (board[2] != "O"):
        return 2
    if (board[1] == board[2] == "X") and (board[0] != "O"):
        return 0
    if (board[3] == board[4] == "X") and (board[5] != "O"):
        return 5
    if (board[4] == board[5] == "X") and (board[3] != "O"):
        return 3
    if (board[6] == board[7] == "X") and (board[8] != "O"):
        return 8
    if (board[7] == board[8] == "X") and (board[6] != "O"):
        return 6
    if (board[0] == board[3] == "X") and (board[6] != "O"):
        return 6
    if (board[3] == board[6] == "X") and (board[0] != "O"):
        return 0
    if (board[1] == board[4] == "X") and (board[7] != "O"):
        return 7
    if (board[4] == board[7] == "X") and (board[1] != "O"):
        return 1
    if (board[2] == board[5] == "X") and (board[8] != "O"):
        return 8
    if (board[5] == board[8] == "X") and (board[2] != "O"):
        return 2
    if (board[0] == board[4] == "X") and (board[8] != "O"):
        return 8
    if (board[2] == board[4] == "X") and (board[6] != "O"):
        return 6
    if (board[6] == board[4] == "X") and (board[2] != "O"):
        return 2
    if (board[8] == board[4] == "X") and (board[0] != "O"):
        return 0
    if (board[0] == board[2] == "X") and (board[1] != "O"):
        return 1
    if (board[3] == board[5] == "X") and (board[4] != "O"):
        return 4
    if (board[6] == board[8] == "X") and (board[7] != "O"):
        return 7
    if (board[0] == board[6] == "X") and (board[3] != "O"):
        return 3
    if (board[2] == board[8] == "X") and (board[5] != "O"):
        return 5
    if (board[1] == board[7] == "X") and (board[4] != "O"):
        return 4
    if (board[0] == board[8] == "X") and (board[4] != "O"):
        return 4
    if (board[2] == board[6] == "X") and (board[4] != "O"):
        return 4
    
    # check for spots where the cpu is one away from losing
    if (board[0] == board[1] == "O") and (board[2] != "X"):
        return 2
    if (board[1] == board[2] == "O") and (board[0] != "X"):
        return 0
    if (board[3] == board[4] == "O") and (board[5] != "X"):
        return 5
    if (board[4] == board[5] == "O") and (board[3] != "X"):
        return 3
    if (board[6] == board[7] == "O") and (board[8] != "X"):
        return 8
    if (board[7] == board[8] == "O") and (board[6] != "X"):
        return 6
    if (board[0] == board[3] == "O") and (board[6] != "X"):
        return 6
    if (board[3] == board[6] == "O") and (board[0] != "X"):
        return 0
    if (board[1] == board[4] == "O") and (board[7] != "X"):
        return 7
    if (board[4] == board[7] == "O") and (board[1] != "X"):
        return 1
    if (board[2] == board[5] == "O") and (board[8] != "X"):
        return 8
    if (board[5] == board[8] == "O") and (board[2] != "X"):
        return 2
    if (board[0] == board[4] == "O") and (board[8] != "X"):
        return 8
    if (board[2] == board[4] == "O") and (board[6] != "X"):
        return 6
    if (board[6] == board[4] == "O") and (board[2] != "X"):
        return 2
    if (board[8] == board[4] == "O") and (board[0] != "X"):
        return 0
    if (board[0] == board[2] == "O") and (board[1] != "X"):
        return 1
    if (board[3] == board[5] == "O") and (board[4] != "X"):
        return 4
    if (board[6] == board[8] == "O") and (board[7] != "X"):
        return 7
    if (board[0] == board[6] == "O") and (board[3] != "X"):
        return 3
    if (board[2] == board[8] == "O") and (board[5] != "X"):
        return 5
    if (board[1] == board[7] == "O") and (board[4] != "X"):
        return 4
    if (board[0] == board[8] == "O") and (board[4] != "X"):
        return 4
    if (board[2] == board[6] == "O") and (board[4] != "X"):
        return 4

    # choose random spot if board is neutral
    available = []

    for i in range(len(board)):
        if board[i] != "X" and board[i] != "O":
            available.append(i)
    
    return random.choice(available)


def check_for_tie(board):
    """Checks if game is a draw based on current state of board.
    @type board: list
    @rtype: None
    """
    available = 0

    for i in board:
        if i != "O" and i != "X":
            available += 1
    
    if available == 0:
        print("IT'S A TIE!")
        time.sleep(DELAY)
        exit()


def main():
    # declare board with list comprehension
    board = [str(i) for i in range (1, 10)]
    
    # game intro
    print("\n~~~~~~~~~~~~~~~~~~~~~~~")
    print("Welcome to Tic Tac Toe!")
    print("~~~~~~~~~~~~~~~~~~~~~~~")
    time.sleep(DELAY)
    print()
    print()
    show_board(board)
    time.sleep(DELAY)
    print("CPU plays first.\n\n")
    time.sleep(DELAY)

    # cpu chooses random spot out of set of corner spots and center
    options = [0, 2, 4, 6, 8]
    board[random.choice(options)] = "X"

    # get player move
    show_board(board)
    p_move = get_p_move(board)
    board[p_move] = "O"
    time.sleep(DELAY)
    print()
    show_board(board)
    time.sleep(DELAY)
    print("CPU move:\n\n")
    time.sleep(DELAY)

    # cpu chooses spot out of remaining corner and center spots
    new_options = []
    for option in options:
        if not(board[option] == "X" or board[option] == "O"):
            new_options.append(option)
    board[random.choice(new_options)] = "X"
    show_board(board)

    # game mainloop
    while True:
        # get player move
        p_move = get_p_move(board)
        board[p_move] = "O"
        time.sleep(DELAY)
        print()
        show_board(board)

        # check if game is either won or tied
        check_for_win(board)
        check_for_tie(board)

        # cpu makes strategic move
        time.sleep(DELAY)
        print("CPU move:\n\n")
        time.sleep(DELAY)
        cpu_move = cpu_strategic_move(board)
        board[cpu_move] = "X"
        time.sleep(DELAY)
        print()
        show_board(board)

        # check if game is either won or tied
        check_for_win(board)
        check_for_tie(board)


if __name__ == '__main__':
    main()

File: java_programs/Tictactoe.java
/**
 * 
 */

import java.util.Random;
import java.util.Scanner;
import java.util.Arrays;
import java.util.ArrayList;

/**
 * @author Satvick Acharya
 *
 */
public class Tictactoe {

	public static void main(String[] args) {
		String[] board = {"1", "2", "3", "4", "5", "6", "7", "8", "9"};
		
		System.out.println("\n~~~~~~~~~~~~~~~~~~~~~~~");
		System.out.println("Welcome to Tic Tac Toe!");
		System.out.println("~~~~~~~~~~~~~~~~~~~~~~~\n\n");
		showBoard(board);
		
		boolean cpu_is_first = whoGoesFirst();

		if (cpu_is_first) {
			System.out.println("CPU move:\n\n");
			int cpu_move = cpuStrategicMove(board);
			board[cpu_move] = "X";
			showBoard(board);
		}

		while (true) {
			int p_move = getPMove(board);
			board[p_move] = "O";
			System.out.println();
			showBoard(board);

			checkIfGameOver(board);

			System.out.println("CPU move:\n\n");
			int cpu_move = cpuStrategicMove(board);
			board[cpu_move] = "X";
			System.out.println();
			showBoard(board);

			checkIfGameOver(board);			
		}
	}
	
	public static void showBoard(String[] board) {
		System.out.println("-------------");
		System.out.println("| " + board[0] + " | " + board[1] + " | " + board[2] + " |");
		System.out.println("|---+---+---|");
		System.out.println("| " + board[3] + " | " + board[4] + " | " + board[5] + " |");
		System.out.println("|---+---+---|");
		System.out.println("| " + board[6] + " | " + board[7] + " | " + board[8] + " |");
		System.out.println("-------------\n\n");
	}

	public static boolean whoGoesFirst() {
		Scanner getUserPreference = new Scanner(System.in);
		System.out.println("Who should go first? (C)PU or (P)layer: ");
		String preference = getUserPreference.nextLine();

		if (preference.equals("C") || preference.equals("c")) {
			return true;
		} else {
			return false;
		}
	}
	
	public static int getPMove(String[] board) {
		ArrayList<Integer> available = new ArrayList<Integer>();
		
		for (int i = 0; i < board.length; i++) {
			if (!(board[i].equals("X") || board[i].equals("O")))
				available.add(i);
		}
		
		Scanner getPMove = new Scanner(System.in);
		System.out.println("Your move: ");
		int move = getPMove.nextInt();

		int rmove = move - 1;
		return rmove;
	}

	public static void checkIfGameOver(String[] board) {
		String spam = checkForWin(board);

		if (spam.equals("X")) {
			System.out.println("CPU WINS!");
			System.exit(0);
		}
		if (spam.equals("O")) {
			System.out.println("YOU WIN!");
			System.exit(0);
		}

		checkForTie(board);
	}

	public static void checkForTie(String[] board) {
		int available = 0;

		for (String i : board) {
			if (!(i.equals("O") || i.equals("X")))
				available++;
		}

		if (available == 0) {
			System.out.println("IT'S A TIE!");
			System.exit(0);
		}
	}
	
	public static String checkForWin(String[] board) {
		if (board[0].equals(board[1]) && board[0].equals(board[2]))
			return board[0];
		else if (board[3].equals(board[4]) && board[3].equals(board[5]))
			return board[3];
		else if (board[6].equals(board[7]) && board[6].equals(board[8]))
			return board[6];
		else if (board[0].equals(board[3]) && board[0].equals(board[6]))
			return board[0];
		else if (board[1].equals(board[4]) && board[1].equals(board[7]))
			return board[1];
		else if (board[2].equals(board[5]) && board[2].equals(board[8]))
			return board[2];
		else if (board[0].equals(board[4]) && board[0].equals(board[8]))
			return board[0];
		else if (board[2].equals(board[4]) && board[2].equals(board[6]))
			return board[2];
		else
			return "false";
	}

	public static int cpuStrategicMove(String[] board) {
	    ArrayList<Integer> possible_moves = new ArrayList<Integer>();

	    for (int spot = 0; spot < board.length; spot++) {
			if (!(board[spot] == "X" || board[spot] == "O"))
				possible_moves.add(spot);
		}

		String[] options = {"X", "O"};

		for (String letter : options) {
			for (int move : possible_moves) {
				String[] board_copy = board.clone();
				board_copy[move] = letter;
				if (checkForWin(board_copy) != "false")
					return move;
			}
		}

		ArrayList<Integer> good_spots = new ArrayList<Integer>();
		int[] good_options = {0, 2, 4, 6, 8};

		for (int i : possible_moves) {
			if (Arrays.asList(good_options).contains(i))
				good_spots.add(i);
		}

		if (good_spots.contains(4))
			return 4;

		if (good_spots.size() > 0) {
			Random rand = new Random();
			int choice = rand.nextInt(good_spots.size());
			return good_spots.get(choice);
		}

		ArrayList<Integer> edge_spots = new ArrayList<Integer>();
		int[] edge_options = {1, 3, 5, 7};

		for (int i : possible_moves) {
			if (Arrays.asList(edge_options).contains(i))
				edge_spots.add(i);
		}

		if (edge_spots.size() > 0) {
			Random rand = new Random();
			int choice = rand.nextInt(edge_spots.size());
			return edge_spots.get(choice);
		}

		return 0;
	}

}
